### DISCUSSION TITLE
Cumulus ignite runtime error

### DISCUSSION THREAD

**Discussion by michelredondo** [srl-labs/containerlab#1457] [Q&A] [answered]
Hi,
while running a clab topo with cvx nodes (ignite runtime), this error appears:

> Jul 04 17:14:31 server1.dcs3.net 7de2df952da4[1279]: INFO[0002] Attaching NIC vm_swp6 (hwaddr aa:c1:ab:18:9e:45) at index 9
> Jul 04 17:14:31 server1.dcs3.net 7de2df952da4[1279]: INFO[0002] Attaching NIC vm_swp7 (hwaddr aa:c1:ab:38:0c:1a) at index 10
> Jul 04 17:14:31 server1.dcs3.net 7de2df952da4[1279]: INFO[0002] Attaching NIC vm_swp8 (hwaddr aa:c1:ab:11:ef:b7) at index 11
> Jul 04 17:14:31 server1.dcs3.net 7de2df952da4[1279]: INFO[0002] Attaching NIC vm_swp9 (hwaddr aa:c1:ab:1b:19:64) at index 12
> Jul 04 17:14:31 server1.dcs3.net 7de2df952da4[1279]: ERRO[0002] Starting instance: [PUT /actions][400] createSyncActionBadRequest  &{FaultMessage:Cannot initialize a MMIO Network Device or add a device to the MMIO Bus. no more IRQs are available}
> Jul 04 17:14:32 server1.dcs3.net 7de2df952da4[1279]: FATA[0002] runtime error for VM "18192e8f02497640": failed to start machine: [PUT /actions][400] createSyncActionBadRequest  &{FaultMessage:Cannot initialize a MMIO Network Device or add a device to the MMIO Bus. no more IRQs are available}

Is there any limit on the number of interfaces supported on the ignite runtime?

paging @networkop

Thanks!

**Comment by networkop** [comment, marked as answer]
AFAIK, there's no ignite-specific limit. Sounds like some hardware or OS limitation. what sort of environment are you running on?

**Reply by michelredondo** [reply]
clab 0.42, running on Linux Centos kernel 5.4.240-1
Docker 23.0.3
HW: dual Intel Xeon Gold 6138

I have just found this:
https://github.com/firecracker-microvm/firecracker/issues/1268

which states that initially Firecracker had support for 11 devices and currently supports up to 19.

In my case, my cvx device has 12 and it's not booting so I understand  we are in an older firecracker release. Is this the case?

Thanks!

**Reply by networkop** [reply]
oh wow, good find @michelredondo ! I've updated my ignite fork with the FC version that contains the fix (11->19 devices).
https://github.com/networkop/ignite/commit/a07126b944d8ceb4fb34baad5d68831b24845db4
If you're interested in testing, you can rebuild containerlab, updating go.mod to point at the latest commit in networkop/ignite.

**Reply by networkop** [reply]
[for the record] the earliest FC version with the fix: https://github.com/firecracker-microvm/firecracker/blob/v0.24.0/src/arch/src/x86_64/layout.rs

**Reply by michelredondo** [reply]
I have rebuild it but I'm still getting the same error. Not sure if there is something else I have to modify.
Thanks @networkop

**Reply by networkop** [reply]
oh yeah, you'd need to build a new ignite-spawn image. this is where the firecracker binary sits. however I don't remember where it's referenced from. would need to dig a bit more when I have time

**Reply by michelredondo** [reply]
ok, thanks. Will look into it.

**Reply by michelredondo** [reply]
I managed to build the ignite-spawn, tested it and now it works:

```
Jul 06 09:34:14 server1.dcs3.net e7c8efaf578c[32509]: INFO[0008] Attaching NIC vm_swp8 (hwaddr aa:c1:ab:a8:19:d7) at index 11
Jul 06 09:34:14 server1.dcs3.net e7c8efaf578c[32509]: INFO[0008] Attaching NIC vm_swp9 (hwaddr aa:c1:ab:e2:f4:00) at index 12
Jul 06 09:34:14 server1.dcs3.net e7c8efaf578c[32509]: INFO[0008] startInstance successful: [PUT /actions][204] createSyncActionNoContent
```
I only had to slightly modify the Dockerfile because release 0.24 Firecracker release does not untar to a "release" directory:

```
RUN mkdir -p release-${FIRECRACKER_VERSION}
RUN wget -qO- https://github.com/firecracker-microvm/firecracker/releases/download/${FIRECRACKER_VERSION}/firecracker-${FIRECRACKER_VERSION}${FIRECRACKER_ARCH_SUFFIX}.tgz | tar -xvz -C release-${FIRECRACKER_VERSION} && \
    mv release-${FIRECRACKER_VERSION}/firecracker-${FIRECRACKER_VERSION}${FIRECRACKER_ARCH_SUFFIX} /usr/local/bin/firecracker && \
    rm -r release-${FIRECRACKER_VERSION}
```
Thanks for your support @networkop !!

**Reply by networkop** [reply]
@michelredondo do you mind upstreaming these changes? I think it wouldn't hurt to have this available for everyone

**Reply by michelredondo** [reply]
Done!!

**Reply by networkop** [reply]
Thanks @michelredondo. Do I now need to build an image and push it to networkop/ignite:dev ?

**Reply by michelredondo** [reply]
yes, that's my understanding. Clab will pull it:
https://github.com/srl-labs/containerlab/blob/582f76ad9773ca94e0e5e9193b8cd1e0ed5cc838/nodes/cvx/cvx.go#L19

**Reply by networkop** [reply]
all done from my side 👍

**Reply by michelredondo** [reply]
tested and working! 

The additional steps for users that already have used ignite is to delete the existing docker image (docker image rm networkop/ignite:dev) and also delete firecracker folder (rm -rf /var/lib/firecracker/) so image import is triggered with new topology deployment.

paging @hellt in case he considers that additional knobs are needed.

Thanks @networkop
